src/notation/internal/notationproject.cpp

```diff
@@ -48,31 +55,102 @@ mu::Ret NotationProject::load(const io::path& path, const io::path& stylePath, b
     TRACEFUNC;
 
     std::string syffix = io::syffix(path);
+    if (syffix != "mscz" && syffix != "mscx") {
+        return doImport(path, stylePath, forceMode);
+    }
 
-    //! NOTE For "mscz", "mscx" see MsczNotationReader
-    //! for others see readers in importexport module
-    auto reader = readers()->reader(syffix);
-    if (!reader) {
-        return make_ret(notation::Err::FileUnknownType, path);
+    QByteArray msczData;
+    if (syffix == "mscz") {
+        RetVal<QByteArray> rv = fileSystem()->readFile(path);
+        if (!rv.ret) {
+            return rv.ret;
+        }
+        msczData = rv.val;
+    } else if (syffix == "mscx") {
+        Ms::Score::FileError err = engraving::compat::mscxToMscz(path.toQString(), &msczData);
+        if (err != Ms::Score::FileError::FILE_NO_ERROR) {
+            return scoreFileErrorToRet(err, path);
+        }
+    } else {
+        UNREACHABLE;
+        return make_ret(Ret::Code::InternalError);
     }
 
-    return load(path, stylePath, reader, forceMode);
+    QBuffer msczBuf(&msczData);
+    MsczReader reader(&msczBuf);
+    reader.setFilePath(path.toQString());
+    reader.open();
+
+    return doLoad(reader, stylePath, forceMode);
 }
 
-mu::Ret NotationProject::load(const io::path& path, const io::path& stylePath, const INotationReaderPtr& reader, bool forceMode)
+mu::Ret NotationProject::doLoad(engraving::MsczReader& reader, const io::path& stylePath, bool forceMode)
 {
+    // Create new engraving project
+    EngravingProjectPtr project = EngravingProject::create();
+    project->setPath(reader.filePath());
+
+    // Load engraving project
+    engraving::Err err = project->loadMscz(reader, forceMode);
+    if (err != engraving::Err::NoError) {
+        return make_ret(err);
+    }
+
+    // Setup master score
+    err = project->setupMasterScore();
+    if (err != engraving::Err::NoError) {
+        return make_ret(err);
+    }
+
+    // Load style if present
+    if (!stylePath.empty()) {
+        project->masterScore()->loadStyle(stylePath.toQString());
+        if (!Ms::MScore::lastError.isEmpty()) {
+            LOGE() << Ms::MScore::lastError;
+        }
+    }
+
+    // Load other stuff from the project file
+    ProjectAudioSettingsPtr audioSettings = std::shared_ptr<ProjectAudioSettings>(new ProjectAudioSettings());
+    Ret ret = audioSettings->read(reader);
+    if (!ret) {
+        return ret;
+    }
+
+    // Set current if all success
+    m_engravingProject = project;
+
+    m_masterNotation = std::shared_ptr<MasterNotation>(new MasterNotation());
```

Comment:
It's a bit confusing, why would we need to create MasterNotation and ProjectAudioSettings twice.The first time in constructor of NotationProject, the second time here.

Reply:
In the constructor, I create these objects to ensure their existence, that is, I want to "guarantee" that the project always returns always valid objects, so as not to check for null in many places.

In loading, I create new objects to ensure that no state remains from the old object and to indicate that these are objects of the new object.

So, a double creation is obtained. It can be optimized, but I don't think it's necessary.